src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py

```diff
@@ -140,30 +135,40 @@ def _attn(self, query, key, value, attention_mask=None, head_mask=None):
         if self.scale_attn_weights:
             scale_factor /= self.head_dim**0.5
 
-        # MQA: (b, sq, nh * hs)
-        # MHA: (b, nh, sq, hs)
+        # MQA models: (batch_size, query_length, num_heads * head_dim)
+        # MHA models: (batch_size, num_heads, query_length, head_dim)
         query_shape = query.shape
         batch_size = query_shape[0]
         key_length = key.size(-1)
         if self.multi_query:
-            # (b, sq, nh, hs) x (b, hs, sk) -> (b, sq, nh, sk)
+            # (batch_size, query_length, num_heads, head_dim) x (batch_size, head_dim, key_length)
+            # -> (batch_size, query_length, num_heads, key_length)
             query_length = query_shape[1]
             attn_shape = (batch_size, query_length, self.num_heads, key_length)
             attn_view = (batch_size, query_length * self.num_heads, key_length)
             # No copy needed for MQA 2, or when layer_past is provided.
             query = query.reshape(batch_size, query_length * self.num_heads, self.head_dim)
         else:
-            # (b, nh, sq, hs) x (b, nh, hs, sk) -> (b, nh, sq, sk)
+            # (batch_size, num_heads, query_length, head_dim) x (batch_size, num_heads, head_dim, key_length)
+            # -> (batch_size, num_heads, query_length, key_length)
             query_length = query_shape[2]
             attn_shape = (batch_size, self.num_heads, query_length, key_length)
             attn_view = (batch_size * self.num_heads, query_length, key_length)
             # Always copies
             query = query.reshape(batch_size * self.num_heads, query_length, self.head_dim)
             # No copy when layer_past is provided.
             key = key.reshape(batch_size * self.num_heads, self.head_dim, key_length)
-        attn_weights = torch.baddbmm(
-            torch.empty(attn_view, dtype=query.dtype, device=query.device), query, key, beta=0, alpha=scale_factor
-        ).view(attn_shape)
+
+        # check if CPU:
+        if query.device == torch.device("cpu"):
+            # this seemed to be needed - on CPU only: check https://github.com/huggingface/transformers/pull/22575/files#r1159858870
+            zeros = torch.zeros(attn_view, dtype=query.dtype, device=query.device)
+            attn_weights = torch.baddbmm(zeros, query, key, beta=1, alpha=scale_factor).view(attn_shape)
```

Comment:
I know it doesn't really matter, but shouldn't this also have beta=0?

Reply:
I tried it with beta=0 and still got the issue .. maybe torch complains if you multiply 0 tensor with 0 ..

Reply:
That should be the least problematic operation here...
The point is baddbmm shouldn't even read the input when beta=0. Also do you have some code to reproduce this? On my machine there is no problem
```
>>> a=torch.full([50, 50, 50], torch.nan)
>>> b=torch.randn([50, 50, 50])
>>> c=torch.randn([50, 50, 50])
>>> d=torch.full([50, 50, 50], 0)
>>> torch.baddbmm(a,b,c, beta=1, alpha=5).sum()
tensor(nan)
>>> torch.baddbmm(a,b,c, beta=0, alpha=5).sum()
tensor(-9915.6406)
>>> torch.baddbmm(d,b,c, beta=0, alpha=5).sum()
tensor(-9915.6406)
>>> torch.baddbmm(d,b,c, beta=1, alpha=5).sum()
tensor(-9915.6406)
>>> (5*torch.bmm(b,c)).sum()
tensor(-9915.6406)
```

Reply:
Now running on a A100:
```

import torch

s=32
dtype=torch.float32


a=torch.full([s, s, s], torch.nan, dtype=dtype)
b=torch.randn([s, s, s], dtype=dtype)
c=torch.randn([s, s, s], dtype=dtype)
d=torch.zeros([s, s, s], dtype=dtype)

y0=torch.baddbmm(a,b,c, beta=1, alpha=5)
y1=torch.baddbmm(a,b,c, beta=0, alpha=5)
y2=torch.baddbmm(d,b,c, beta=0, alpha=5)
y3=torch.baddbmm(d,b,c, beta=1, alpha=5)
y4=torch.bmm(b,c)*5
y5=torch.matmul(b,c)*5
yy=[y0,y1,y2,y3,y4,y5]

aa=a.cuda()
bb=b.cuda()
cc=c.cuda()
dd=d.cuda()

z0=torch.baddbmm(aa,bb,cc, beta=1, alpha=5).cpu()
z1=torch.baddbmm(aa,bb,cc, beta=0, alpha=5).cpu()
z2=torch.baddbmm(dd,bb,cc, beta=0, alpha=5).cpu()
z3=torch.baddbmm(dd,bb,cc, beta=1, alpha=5).cpu()
z4=(torch.bmm(bb,cc)*5).cpu()
z5=(torch.matmul(bb,cc)*5).cpu()
zz=[z0,z1,z2,z3,z4,z5]

print([(z-y).std() for y,z in zip(yy,zz)])

print([(y-yy[1]).std() for y in yy])
print([(z-zz[1]).std() for z in zz])
```

```
>>> print([(z-y).std() for y,z in zip(yy,zz)])
[tensor(nan), tensor(0.0084), tensor(0.0084), tensor(0.0084), tensor(0.0084), tensor(0.0084)]
>>> 
>>> print([(y-yy[1]).std() for y in yy])
[tensor(nan), tensor(0.), tensor(0.), tensor(0.), tensor(4.2393e-06), tensor(4.2393e-06)]
>>> print([(z-zz[1]).std() for z in zz])
[tensor(nan), tensor(0.), tensor(0.), tensor(0.), tensor(0.), tensor(0.)]
>>>  
```
From this output it looks like the output is indeed different on CPU vs GPU, but that happens for every single kind of matrix multiplication so there is nothing we can do about it... In general I don't think we should expect the numerically equal outputs on different devices

Reply:
I could reproduce through a common test, what I did was to replace the following block with:
```python
if False:
            # this seemed to be needed - on CPU only: check https://github.com/huggingface/transformers/pull/22575/files#r1159858870
            zeros = torch.zeros(attn_view, dtype=query.dtype, device=query.device)
            attn_weights = torch.baddbmm(zeros, query, key, beta=1, alpha=scale_factor).view(attn_shape)
 else:
    # We do the standard operation on GPU for faster inference
    attn_weights = torch.baddbmm(
        torch.empty(attn_view, device=query.device, dtype=query.dtype), query, key, beta=0, alpha=scale_factor
    ).view(attn_shape)

    if attn_weights.isnan().any():
        print()
```
And run 
```bash
pytest tests/models/gpt_bigcode/test_modeling_gpt_bigcode.py::GPTBigCodeMQAModelTest::test_beam_sample_generate
```
and put a breakpoint right on the print. Strangely I couldn't reproduce in a small snippet. Here is what I have tried:
```python
import torch

N_EXPERIMENTS = 10000
device = torch.device("cpu")

batch_size=8
q_len=4
k_len=5
hidden_dim=8
dtype=torch.float32

attn_view =(batch_size, q_len, k_len)
attn_shape = (batch_size, 1, q_len, k_len)

for _ in range(N_EXPERIMENTS):
    query = torch.randn(batch_size, q_len, hidden_dim, device=device, dtype=dtype)
    key = torch.randn(batch_size, hidden_dim, k_len, device=device, dtype=dtype)

    a = torch.empty(attn_view, device=query.device, dtype=query.dtype)
    out = torch.baddbmm(a, query, key, beta=0, alpha=0.356).view(attn_shape)

    if out.isnan().any():
        print(out.isnan().any())
```

Reply:
I did a bit more investigating, from what I could find that's a semi-random bug that only manifests itself when it feels like it so it's not easy to reproduce. I was able to get it consistently with
```
import torch
s=7
dtype=torch.float32
aa=[]

n=[]
for i in range(10000):
    a=torch.full([s,s,s],torch.nan, dtype=dtype)
    b=torch.randn([s,s,s], dtype=dtype)
    c=torch.randn([s,s,s], dtype=dtype)
    y=torch.baddbmm(a,b,c, beta=0, alpha=5)
    aa+=[a,b,c]
    n.append(y.isnan().float().mean().item())

>>> print(torch.mean(torch.tensor(n)).item())
0.2447994202375412
```
For some reason I never get any nan with s>=8 and rarely when not accumulating tensors in the list (aka reusing memory addreses).

Reply:
Anyway, it looks like that's a known bug https://github.com/pytorch/pytorch/issues/96037. It's been fixed in https://github.com/pytorch/pytorch/pull/96086 but only for the next release of pytorch. So for now we should leave the zero, I'll just simplify that code and add a reference to the issue in the comment. (In the future it could be updated to only set to zero for torch version <=2.0.0)

Reply:
Thanks a mile for investigating!